src/main.rs

```diff
@@ -33,213 +29,274 @@ fn main() -> anyhow::Result<()> {
+    let metadata = MetadataCommand::new()
+        .cargo_path(&cargo_path)
+        .manifest_path(&opts.manifest_path)
+        .no_deps()
+        .exec()?;
+
+    let focus_package = match opts.package {
+        Some(name) => metadata
+            .packages
+            .iter()
+            .find(|p| p.name == name)
+            .with_context(|| format!("Package '{}' is not found", name))?,
+        None if metadata.packages.len() == 1 => &metadata.packages[0],
+        None => anyhow::bail!("Multiple packages found"),
+    };
 
-    compile_opts.spec = Packages::Packages(vec![package.clone()]);
+    let focus_artifact = match opts.focus {
+        Some(focus) => focus,
+        None => match focus_package.targets.len() {
+            0 => anyhow::bail!("No targets found"),
+            1 => opts::Focus::try_from(&focus_package.targets[0])?,
+            _ => anyhow::bail!("Multiple targets found"),
+        },
+    };
 
-    if let Some(focus) = &opts.focus {
-        compile_opts.filter = CompileFilter::from(focus.clone());
-    }
-    compile_opts.cli_features = opts.cli_features.try_into()?;
-    compile_opts.build_config.requested_kinds = vec![kind];
-    compile_opts.build_config.requested_profile = InternedString::from(&opts.compile_mode);
-    compile_opts.build_config.force_rebuild = opts.force_rebuild;
+    let mut cargo_child = {
+        use std::ffi::OsStr;
+
+        let mut cmd = std::process::Command::new(&cargo_path);
+
+        // Cargo flags.
+        cmd.arg("rustc")
+            // General.
+            .args([
+                "--message-format=json",
+                "--color",
+                if opts.format.color { "always" } else { "never" },
+            ])
+            .args(std::iter::repeat("-v").take(opts.format.verbosity as usize))
+            // Workspace location.
+            .arg("--manifest-path")
+            .arg(opts.manifest_path)
+            // Artifact selectors.
+            .args(["--package", &focus_package.name])
+            .args(focus_artifact.as_cargo_args())
+            // Compile options.
+            .args(opts.dry.then_some("--dry"))
+            .args(opts.frozen.then_some("--frozen"))
+            .args(opts.locked.then_some("--locked"))
+            .args(opts.offline.then_some("--offline"))
+            .args(opts.target.iter().flat_map(|t| ["--target", t]))
+            .args(
+                opts.target_dir
+                    .iter()
+                    .flat_map(|t| [OsStr::new("--target-dir"), t.as_ref()]),
+            )
+            .args(
+                opts.cli_features
+                    .no_default_features
+                    .then_some("--no-default-features"),
+            )
+            .args(opts.cli_features.all_features.then_some("--all-features"))
+            .args(
+                opts.cli_features
+                    .feature
+                    .iter()
+                    .flat_map(|feat| ["--feature", feat]),
```

Comment:
My fault for picking a wrong name
```suggestion
                    .flat_map(|feat| ["--features", feat]),
```